include/ignition/rendering/BoundingBoxCamera.hh

```diff
@@ -0,0 +1,162 @@
+      ignition::math::Vector3d center;
+
+      /// \brief Size of the box (width, height, depth), depth = 0 in 2D boxes
+      ignition::math::Vector3d size;
+
+      /// \brief Orientation of the 3D box in camera coord.
+      /// The 2D boxes are axis aligned (orientation = 0)
+      ignition::math::Quaterniond orientation;
+
+      /// \brief Label of the annotated object inside the box
+      uint32_t label;
+
+      /// \brief Constructor
+      /// \param[in] _type Bounding Box Type
+      /// \param[in] _label Label for the object in the bounding box
+      BoundingBox(ignition::rendering::BoundingBoxType _type, uint32_t _label)
+      {
+        this->type = _type;
+        this->label = _label;
+      }
+
+      /// \brief Get the vertices of the 3D bounding box. If the bounding box
+      /// type isn't 3D, an empty vector is returned
+      /// \return The vertices of the 3D bounding box
+      std::vector<math::Vector3d> Vertices() const
+      {
+        if (this->type != BoundingBoxType::BBT_BOX3D)
+          return {};
+
+        /*
+          The numbers listed here are the corresponding indices in the vector
+          that is returned by this method
+
+            1 -------- 0
+           /|         /|
+          2 -------- 3 .
+          | |        | |
+          . 5 -------- 4
+          |/         |/
+          6 -------- 7
+        */
+
+        std::vector<math::Vector3d> vertices;
+        auto w = this->size.X();
+        auto h = this->size.Y();
+        auto l = this->size.Z();
+
+        // 8 vertices | box corners
+        vertices.push_back(math::Vector3d(w/2, h/2, l/2));
+        vertices.push_back(math::Vector3d(-w/2, h/2, l/2));
+        vertices.push_back(math::Vector3d(-w/2, h/2, -l/2));
+        vertices.push_back(math::Vector3d(w/2, h/2, -l/2));
+        vertices.push_back(math::Vector3d(w/2, -h/2, l/2));
+        vertices.push_back(math::Vector3d(-w/2, -h/2, l/2));
+        vertices.push_back(math::Vector3d(-w/2, -h/2, -l/2));
+        vertices.push_back(math::Vector3d(w/2, -h/2, -l/2));
+
+        // Transform
+        for (auto &vertex : vertices)
+        {
+          vertex = this->orientation * vertex + this->center;
+        }
+
+        return vertices;
+      }
+    };
+
+    inline namespace IGNITION_RENDERING_VERSION_NAMESPACE {
+    /// \class BoundingBoxCamera BoundingBoxCamera.hh
+    /// ignition/rendering/BoundingBoxCamera.hh
+    /// \brief Poseable BoundingBox camera used for rendering bounding boxes of
+    /// objects in the scene.
+    class IGNITION_RENDERING_VISIBLE BoundingBoxCamera :
+      public virtual Camera
+    {
+      /// \brief Destructor
+      public: virtual ~BoundingBoxCamera() { }
+
+      /// \brief Get the BoundingBox data
+      /// \return Buffer of bounding boxes info (label, minX, minY, maxX, maxY)
+      public: virtual const std::vector<BoundingBox> &BoundingBoxData()
+              const = 0;
```

Comment:
Would it be better to return a copy here instead of a reference in case users want to do things with the bounding box data? I also wonder if a copy is better so that if the bounding box data changes frequently, users don't have to worry about this since they have a copy

Reply:
const ref is fine, the caller can always make a copy if they want to, and this way we're not forcing the copy.

Reply:
I think this API gives user a cons ref and they also have the option to make a copy of it if they like right? So if they plan to store the vector and do something with the data, they'll just need to do an addition call to manually copy it, e.g. with std::copy

Reply:
Sounds good, thanks for the feedback :+1: I will leave it as-is, returning a const ref